File: frontend/src/api/schemas/report.ts
import { z } from 'zod'
import { NoteAuthorSchema } from './note'

export const ReportReasonSchema = z.enum([
  'spam',
  'harassment',
  'sexual',
  'illegal',
  'misinfo',
  'infringement',
  'other',
])
export type ReportReason = z.infer<typeof ReportReasonSchema>

/** Reason options with Chinese labels, in display order. */
export const REPORT_REASONS: ReadonlyArray<{ value: ReportReason; label: string }> = [
  { value: 'spam', label: '垃圾广告' },
  { value: 'harassment', label: '辱骂骚扰' },
  { value: 'sexual', label: '色情低俗' },
  { value: 'illegal', label: '违法有害' },
  { value: 'misinfo', label: '不实信息' },
  { value: 'infringement', label: '侵权' },
  { value: 'other', label: '其他' },
]

export const reasonLabel = (v: string): string =>
  REPORT_REASONS.find((r) => r.value === v)?.label ?? v

/** Mirror of backend ReportOut. */
export const ReportSchema = z.object({
  id: z.string(),
  targetType: z.string(),
  targetNoteId: z.string().nullish(),
  targetCommentId: z.string().nullish(),
  targetSnapshot: z.string().default(''),
  reason: z.string(),
  description: z.string().nullish(),
  status: z.string(),
  aiLabel: z.string().nullish(),
  aiConfidence: z.number().nullish(),
  aiReason: z.string().nullish(),
  resolutionAction: z.string().nullish(),
  resolutionComment: z.string().nullish(),
  resolvedBySid: z.string().nullish(),
  resolvedAt: z.string().nullish(),
  createdAt: z.string(),
  updatedAt: z.string(),
  reporter: NoteAuthorSchema.nullish(),
})
export type Report = z.infer<typeof ReportSchema>
export const ReportListSchema = z.array(ReportSchema)

/** Mirror of backend BlockOut. */
export const BlockSchema = z.object({
  user: NoteAuthorSchema,
  createdAt: z.string(),
})
export type Block = z.infer<typeof BlockSchema>
export const BlockListSchema = z.array(BlockSchema)

export type ReportCreateIn = {
  targetType: 'note' | 'comment'
  targetId: string
  reason: ReportReason
  description?: string
}
export type ReportResolveIn = { action: 'hide' | 'delete' | 'dismiss'; comment?: string }

File: backend/app/schemas/report.py
"""Governance schemas — reports, moderation resolutions, blocks."""
from __future__ import annotations

from typing import Literal

from pydantic import Field

from app.schemas._base import CamelModel, UtcDateTime
from app.schemas.note import NoteAuthorOut

ReportReason = Literal[
    "spam",
    "harassment",
    "sexual",
    "illegal",
    "misinfo",
    "infringement",
    "other",
]
ReportTargetType = Literal["note", "comment"]
ResolutionAction = Literal["hide", "delete", "dismiss"]


class ReportCreateIn(CamelModel):
    target_type: ReportTargetType
    target_id: str
    reason: ReportReason
    description: str | None = Field(default=None, max_length=1000)


class ReportResolveIn(CamelModel):
    action: ResolutionAction
    comment: str | None = Field(default=None, max_length=1000)


class ReportOut(CamelModel):
    id: str
    target_type: str
    target_note_id: str | None = None
    target_comment_id: str | None = None
    target_snapshot: str = ""
    reason: str
    description: str | None = None
    status: str
    ai_label: str | None = None
    ai_confidence: float | None = None
    ai_reason: str | None = None
    resolution_action: str | None = None
    resolution_comment: str | None = None
    resolved_by_sid: str | None = None
    resolved_at: UtcDateTime | None = None
    created_at: UtcDateTime
    updated_at: UtcDateTime
    reporter: NoteAuthorOut | None = None


class BlockOut(CamelModel):
    user: NoteAuthorOut
    created_at: UtcDateTime
